downloader/src/main/kotlin/VcsHost.kt

```diff
@@ -0,0 +1,172 @@
+ */
+class Bitbucket(url: String) : VcsHost(url) {
+    override val hostname = "bitbucket.org"
+    override val isGitOnly = false
+
+    override fun toVcsInfo(): VcsInfo? {
+        if (uri == null) return null
+        var url = uri.scheme + "://" + uri.authority
+
+        // Append the first two path components that denote the user and project to the base URL.
+        val pathIterator = Paths.get(uri.path).iterator()
+        if (pathIterator.hasNext()) {
+            url += "/${pathIterator.next()}"
+        }
+        if (pathIterator.hasNext()) {
+            url += "/${pathIterator.next()}"
+        }
+
+        var revision = ""
+        var path = ""
+
+        if (pathIterator.hasNext() && pathIterator.next().toString() == "src") {
+            if (pathIterator.hasNext()) {
+                revision = pathIterator.next().toString()
+                path = uri.path.substringAfter(revision).trimStart('/').removeSuffix(".git")
+            }
+        }
+
+        val type = VersionControlSystem.forUrl(url)?.type ?: VcsType.NONE
+        if (type == VcsType.GIT) {
+            url += ".git"
+        }
+
+        return VcsInfo(type, url, revision, path = path)
+    }
+}
+
+private fun gitUrlToVcsInfo(uri: URI): VcsInfo {
+    var url = uri.scheme + "://" + uri.authority
+
+    // Append the first two path components that denote the user and project to the base URL.
+    val pathIterator = Paths.get(uri.path).iterator()
+    if (pathIterator.hasNext()) {
+        url += "/${pathIterator.next()}"
+    }
+    if (pathIterator.hasNext()) {
+        url += "/${pathIterator.next()}"
+
+        if (!url.endsWith(".git")) {
+            url += ".git"
+        }
+    }
+
+    var revision = ""
+    var path = ""
+
+    if (pathIterator.hasNext()) {
+        val extra = pathIterator.next().toString()
+        if (extra in listOf("blob", "tree") && pathIterator.hasNext()) {
+            revision = pathIterator.next().toString()
+            path = uri.path.substringAfter(revision).trimStart('/').removeSuffix(".git")
+        } else {
+            // Just treat all the extra components as a path.
+            path = (sequenceOf(extra) + pathIterator.asSequence()).joinToString("/")
+        }
+    } else {
+        if (uri.hasFragmentRevision()) revision = uri.fragment
+    }
+
+    return VcsInfo(VcsType.GIT, url, revision, path = path)
+}
+
+/**
+ * A class to handle [GitHub][https://github.com/]-specific information.
+ */
+class GitHub(url: String) : VcsHost(url) {
+    override val hostname = "github.com"
+    override val isGitOnly = true
+
+    override fun toVcsInfo() = if (isGitOnly && uri != null) gitUrlToVcsInfo(uri) else null
+}
+
+/**
+ * A class to handle [GitLab][https://gitlab.com/]-specific information.
+ */
+class GitLab(url: String) : VcsHost(url) {
+    override val hostname = "gitlab.com"
+    override val isGitOnly = true
+
+    override fun toVcsInfo() = if (isGitOnly && uri != null) gitUrlToVcsInfo(uri) else null
```

Comment:
```isGitOnly``` is always true.

Reply:
Correct. I did that to emphasize that calling `gitUrlToVcsInfo()` is only valid if the host *only* supports Git.

Reply:
I have a bit doubts that different readers will understand it in the way you intended it, anyhow I'll leave that one up to you to decide..

Reply:
That code is not present anymore in the newest iteration anyway.